### PR TITLE
8236919: Refactor com.sun.tools.javac.main.CommandLine into a reusable module for other JDK tools

### PR DESCRIPTION
**Pull request by sormuras**
This PR copies the `CommandLine.java` file from module `jdk.compiler` (package `com.sun.tools.javac.main`) into the `jdk.internal.opt` module, creating a new package with name `jdk.internal.opt`. That new `jdk.internal.opt` package is then exported to the following modules:
- `jdk.jartool`
- `jdk.jlink`
- `jdk.jpackage`

Now, `jar`, `jlink`, and `jpackage` use a shared `CommandLine` class. In a future commit (presumable for JDK 21)  the original `CommandLine.java` file in `jdk.compiler` can and will be replaced with this new one in `jdk.internal.opt`. Same goes for the `jdk.javadoc` module.

- [x] Keep `CommandLine.java` in `jdk.compiler` module for the time being due to "JDK N-1 rule".
- [x] Keep `CommandLine.java` in `jdk.javadoc` module for the time being due to "JDK N-1 rule".
- [x] Remove `CommandLine.java` from `jdk.jartool` module
- [x] Remove `CommandLine.java` from `jdk.jlink` module
- [x] Remove `CommandLine.java` from `jdk.jpackage` module
- [x] Check for related but renamed(?) usages of `CommandLine.java` in other JDK tools: `jshell`, `jdeps`, `jfr`, ...

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8236919](https://bugs.openjdk.org/browse/JDK-8236919): Refactor com.sun.tools.javac.main.CommandLine into a reusable module for other JDK tools

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11272/head:pull/11272` \
`$ git checkout pull/11272`

Update a local copy of the PR: \
`$ git checkout pull/11272` \
`$ git pull https://git.openjdk.org/jdk pull/11272/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11272`

View PR using the GUI difftool: \
`$ git pr show -t 11272`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11272.diff">https://git.openjdk.org/jdk/pull/11272.diff</a>

</details>

### CODE REVIEW

File: src/jdk.internal.opt/share/classes/jdk/internal/opt/CommandLine.java [hunk 1]
```diff
@@ -50,7 +50,16 @@
  *  This code and its internal interfaces are subject to change or
  *  deletion without notice.</b>
```
Comment (jonathan-gibbons) [inline src/jdk.internal.opt/share/classes/jdk/internal/opt/CommandLine.java:51]: Maybe we can drop this (part of the) comment

**Review by jonathan-gibbons — APPROVED** [review]
I guess git is "confused" by not realizing that the "new" `CommandLine.java` is just a clone of the existing `javac` version. (It is, isn't it ... apart from the package rename?)

**Comment by sormuras** [comment]
Yes. I copied `CommandLine` class from module `jdk.compiler`. In addition to update of the copyright year and the name of the package, I also added small API changes that other tools developed in their copies of the class. Find a diff attached below:

```diff
diff --git a/open/src/jdk.compiler/share/classes/com/sun/tools/javac/main/CommandLine.java b/open/src/jdk.internal.opt/share/classes/jdk/internal/opt/CommandLine.java
index ec6f711f9b..9e4b0c6fe2 100644
--- a/open/src/jdk.compiler/share/classes/com/sun/tools/javac/main/CommandLine.java
+++ b/open/src/jdk.internal.opt/share/classes/jdk/internal/opt/CommandLine.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1999, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1999, 2022, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,17 +23,25 @@
  * questions.
  */
 
-package com.sun.tools.javac.main;
+package jdk.internal.opt;
 
 import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
 import java.io.Reader;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Paths;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.List;
 
+/*
+ * This file was originally a copy of CommandLine.java in
+ * com.sun.tools.javac.main -- and it will be the last.
+ *
+ * Find details at https://bugs.openjdk.org/browse/JDK-8236919
+ */
+
 /**
  * Various utility methods for processing Java tool command line arguments.
  *
@@ -42,7 +50,16 @@ import java.util.List;
  *  This code and its internal interfaces are subject to change or
  *  deletion without notice.</b>
  */
-public class CommandLine {
+public final class CommandLine {
+    /**
+     * Convenient wrapper for the {@code List}-based parse method.
+     *
+     * @see #parse(List)
+     */
+    public static String[] parse(String... args) throws IOException {
+        return parse(List.of(args)).toArray(String[]::new);
+    }
+
     /**
      * Process Win32-style command files for the specified command line
      * arguments and return the resulting arguments. A command file argument
@@ -91,7 +108,7 @@ public class CommandLine {
      * @param args the arguments that may contain @files
      * @return the arguments, with environment variable's content and expansion of @files
      * @throws IOException if there is a problem reading any of the @files
-     * @throws com.sun.tools.javac.main.CommandLine.UnmatchedQuote
+     * @throws CommandLine.UnmatchedQuote
      */
     public static List<String> parse(String envVariable, List<String> args)
             throws IOException, UnmatchedQuote {
@@ -104,8 +121,18 @@ public class CommandLine {
         return newArgs;
     }
 
+    public static void loadCmdFile(InputStream in, List<String> args) throws IOException {
+        Reader reader = new InputStreamReader(in);
+        loadCmdFileAndCloseReader(reader, args);
+    }
+
     private static void loadCmdFile(String name, List<String> args) throws IOException {
-        try (Reader r = Files.newBufferedReader(Paths.get(name), Charset.defaultCharset())) {
+        Reader reader = Files.newBufferedReader(Paths.get(name), Charset.defaultCharset());
+        loadCmdFileAndCloseReader(reader, args);
+    }
+
+    private static void loadCmdFileAndCloseReader(Reader r, List<String> args) throws IOException {
+        try (r) {
             Tokenizer t = new Tokenizer(r);
             String s;
             while ((s = t.nextToken()) != null) {
```

File: src/jdk.internal.opt/share/classes/jdk/internal/opt/CommandLine.java [hunk 2]
```diff
@@ -46,9 +46,7 @@
  * Various utility methods for processing Java tool command line arguments.
  *
  *  <p><b>This is NOT part of any supported API.
- *  If you write code that depends on this, you do so at your own risk.
- *  This code and its internal interfaces are subject to change or
- *  deletion without notice.</b>
+ *  If you write code that depends on this, you do so at your own risk.</b>
```
Comment (jonathan-gibbons) [inline src/jdk.internal.opt/share/classes/jdk/internal/opt/CommandLine.java:49]: I think this entire paragraph can be removed. It is "just" the "scary warning" meant to warn people about using not-really-public classes in javac, before we had a module system for the encapsulation.

**Review by jonathan-gibbons — APPROVED** [review]

**Review by jonathan-gibbons — APPROVED** [review]

**Pull request closed without merging**